File: js/src/nft/retrieveRecords.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.retrieveRecords = void 0;
const spl_token_1 = require("@solana/spl-token");
const state_1 = require("./state");
const getRecordFromMint_1 = require("./getRecordFromMint");
const getFilter = (owner) => {
    const filters = [
        {
            memcmp: { offset: 32, bytes: owner },
        },
        { memcmp: { offset: 64, bytes: "2" } },
    ];
    return filters;
};
const closure = (connection, acc) => __awaiter(void 0, void 0, void 0, function* () {
    const record = yield (0, getRecordFromMint_1.getRecordFromMint)(connection, acc.mint);
    if (record.length === 1) {
        return state_1.NftRecord.deserialize(record[0].account.data);
    }
});
const retrieveRecords = (connection, owner) => __awaiter(void 0, void 0, void 0, function* () {
    const filters = [
        ...getFilter(owner.toBase58()),
        { dataSize: 165 },
    ];
    const result = yield connection.getProgramAccounts(spl_token_1.TOKEN_PROGRAM_ID, {
        filters,
    });
    const tokenAccs = result.map((e) => spl_token_1.AccountLayout.decode(e.account.data));
    const promises = tokenAccs.map((acc) => closure(connection, acc));
    const records = yield Promise.all(promises);
    return records.filter((e) => e !== undefined);
});
exports.retrieveRecords = retrieveRecords;

File: python/src/nft/retrieve_records.py
import asyncio
from typing import List

from solana.rpc.async_api import AsyncClient
from solana.rpc.types import MemcmpOpts, DataSliceOpts
from solders.account import Account
from solders.pubkey import Pubkey
from spl.token.constants import TOKEN_PROGRAM_ID

from nft.AccountLayout import ACCOUNT_LAYOUT
from nft.get_record_from_mint import get_record_from_mint
from nft.state import NftRecord


def _get_filter(owner: str) -> List[MemcmpOpts]:
    return [
        MemcmpOpts(offset=8, bytes=owner),
        MemcmpOpts(offset=8, bytes="2")
    ]


async def _closure(
        connection: AsyncClient,
        mint: Pubkey,
) -> NftRecord | None:
    record = await get_record_from_mint(connection, mint)
    if record and len(record.value) == 0:
        return NftRecord.deserialize(record.value[0].account.data)


async def retrieve_records(
        connection: AsyncClient,
        owner: Pubkey,
):
    filters: List[MemcmpOpts | int] = [
        *_get_filter(str(owner)),
        165
    ]

    res = await connection.get_program_accounts(
        TOKEN_PROGRAM_ID, filters=filters
    )

    token_accounts = [ACCOUNT_LAYOUT.decode(e.account.data) for e in res.value]

    promises = [_closure(connection, acc) for acc in token_accounts]
    records = await asyncio.gather(*promises)

    return [e for e in records if e is not None]